src/libraries/Common/src/System/Security/Cryptography/X509Certificates/X509CertificateLoader.Pkcs12.cs

```diff
@@ -0,0 +1,909 @@
+                {
+                    HResult = CRYPT_E_BAD_DECODE,
+                };
+            }
+        }
+
+        private static void ProcessSafeContents(
+            ReadOnlyMemory<byte> contentData,
+            Pkcs12LoaderLimits loaderLimits,
+            ref int? workRemaining,
+            ref BagState bagState)
+        {
+            AsnValueReader outer = new AsnValueReader(contentData.Span, AsnEncodingRules.BER);
+            AsnValueReader reader = outer.ReadSequence();
+            outer.ThrowIfNotEmpty();
+
+            while (reader.HasData)
+            {
+                SafeBagAsn.Decode(ref reader, contentData, out SafeBagAsn bag);
+
+                if (bag.BagId == Oids.Pkcs12CertBag)
+                {
+                    CertBagAsn certBag = CertBagAsn.Decode(bag.BagValue, AsnEncodingRules.BER);
+
+                    if (certBag.CertId == Oids.Pkcs12X509CertBagType)
+                    {
+                        if (bagState.CertCount >= loaderLimits.MaxCertificates)
+                        {
+                            throw new Pkcs12LoadLimitExceededException(nameof(Pkcs12LoaderLimits.MaxCertificates));
+                        }
+
+                        if (bag.BagAttributes is not null)
+                        {
+                            FilterAttributes(
+                                loaderLimits,
+                                ref bag,
+                                static (limits, oid) => limits.PreserveUnknownAttributes || oid == Oids.LocalKeyId);
+                        }
+
+                        bagState.AddCert(bag);
+                    }
+                }
+                else if (bag.BagId == Oids.Pkcs12KeyBag || bag.BagId == Oids.Pkcs12ShroudedKeyBag)
+                {
+                    if (bagState.KeyCount >= loaderLimits.MaxKeys)
+                    {
+                        throw new Pkcs12LoadLimitExceededException(nameof(Pkcs12LoaderLimits.MaxKeys));
+                    }
+
+                    if (bag.BagId == Oids.Pkcs12ShroudedKeyBag)
+                    {
+                        EncryptedPrivateKeyInfoAsn epki = EncryptedPrivateKeyInfoAsn.Decode(
+                            bag.BagValue,
+                            AsnEncodingRules.BER);
+
+                        int kdfCount = GetKdfCount(epki.EncryptionAlgorithm);
+
+                        if (kdfCount > loaderLimits.IndividualKdfIterationLimit || kdfCount > workRemaining)
+                        {
+                            throw new Pkcs12LoadLimitExceededException(
+                                kdfCount > loaderLimits.IndividualKdfIterationLimit ?
+                                    nameof(Pkcs12LoaderLimits.IndividualKdfIterationLimit) :
+                                    nameof(Pkcs12LoaderLimits.TotalKdfIterationLimit));
+                        }
+
+                        if (workRemaining.HasValue)
+                        {
+                            workRemaining -= kdfCount;
```

Comment:
From this comment I hardened GetKdfCount to not allow a negative return.

Given that this is the only mutation of `workRemaining`, and it provably can't overflow (because the previous if block was if-greater-than-then-throw) I'm not sure that adding checked has value here, other than virtue signalling.